File: SunoFlowApp/Sources/SunoFlow/TranscriptionClient.swift
import Foundation

struct TranscriptionResult: Decodable {
    let raw: String
    let cleaned: String
}

/// Which half of the dictionary an entry belongs to.
///
/// The two behave very differently at dictation time — a spelling is swapped in
/// mechanically, a shorthand value is only substituted where the model judges
/// the speaker was actually giving it — so the UI names the difference rather
/// than showing one undifferentiated list. See docs/CONTRACT.md §corrections.
enum CorrectionKind: String, CaseIterable, Identifiable {
    case correction
    case expansion

    var id: String { rawValue }

    var label: String {
        switch self {
        case .correction: return "Spelling"
        case .expansion:  return "Shorthand"
        }
    }

    var blurb: String {
        switch self {
        case .correction: return "A word dictation mishears, and how you write it."
        case .expansion:  return "Something you say out loud, and the value it stands for."
        }
    }

    var fromPlaceholder: String {
        switch self {
        case .correction: return "What it mishears"
        case .expansion:  return "What you say"
        }
    }

    var toPlaceholder: String {
        switch self {
        case .correction: return "What it should write"
        case .expansion:  return "The value to write instead"
        }
    }
}

struct Correction: Decodable, Identifiable {
    let key: String
    let from: String
    let to: String
    let count: Int
    /// Decoded as a plain String rather than as `CorrectionKind` on purpose: a
    /// raw value the enum does not know makes Swift throw, and a throw here
    /// empties the whole list silently. Absent altogether from a sidecar that
    /// predates the two-kind dictionary, and those only ever held spellings.
    private let kind: String?

    var id: String { key }

    var resolvedKind: CorrectionKind {
        CorrectionKind(rawValue: kind ?? "") ?? .correction
    }
}

private struct CorrectionsResponse: Decodable {
    let corrections: [Correction]
}

/// A single learned pair returned by `POST /learn`. The server does NOT send a
/// `key` field (unlike `GET /corrections`), so this is a lighter struct than
/// `Correction` — decoding `learned` into `[Correction]` fails silently and the
/// learned count always reads 0.
private struct LearnedItem: Decodable {
    let from: String
    let to: String
    let count: Int
}

private struct LearnResponse: Decodable {
    let learned: [LearnedItem]
}

/// Response shape of the cleanup gateway's unauthenticated `/ready` probe.
/// `backend_ok` is true only when the gateway's LLM backend is reachable.
private struct ReadyResponse: Decodable {
    let backend_ok: Bool
}

/// Progress/state of the STT model, reported by the sidecar's `/model/status`.
struct ModelStatus: Decodable {
    let model_present: Bool
    let model_loaded: Bool
    let active: Bool
    let phase: String        // idle | downloading | loading | done | error
    let current_file: String
    let downloaded: Int64
    let file_total: Int64
    let overall_done: Int
    let overall_total: Int
    let error: String
    let model_dir: String
    let model_id: String
}

private struct DownloadStartResponse: Decodable {
    let started: Bool
    let reason: String?
}

enum TranscriptionError: Error {
    case badResponse
    /// This Mac may not dictate. Carries the wording the gateway chose, so the
    /// app shows one consistent explanation, plus a code telling the two very
    /// different reasons apart: `not_entitled` (trial over, subscription
    /// lapsed, device disconnected — fixed on the account page) and
    /// `unreachable` (we could not check — fixed by reconnecting). Calling an
    /// outage a cancelled subscription is a support ticket waiting to happen.
    case notEntitled(code: String, message: String)
}

private func formURLEncoded(_ fields: [String: String]) -> Data {
    var allowed = CharacterSet.alphanumerics
    allowed.insert(charactersIn: "-._~")
    let body = fields.map { key, value in
        let k = key.addingPercentEncoding(withAllowedCharacters: allowed) ?? key
        let v = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
        return "\(k)=\(v)"
    }.joined(separator: "&")
    return body.data(using: .utf8) ?? Data()
}

enum TranscriptionClient {
    static let baseURL = URL(string: "http://127.0.0.1:8765")!

    /// Base URL of the hosted cleanup gateway (transcript polishing service).
    /// The gateway owns the cleanup instruction and LLM backend; the app only
    /// probes its `/ready` for the settings connectivity status.
    static let gatewayURL = URL(string: "https://cleanup.mirrorli.art")!

    static func health(completion: @escaping (Bool) -> Void) {
        var request = URLRequest(url: baseURL.appendingPathComponent("health"))
        request.timeoutInterval = 2
        URLSession.shared.dataTask(with: request) { _, response, _ in
            let ok = (response as? HTTPURLResponse)?.statusCode == 200
            completion(ok)
        }.resume()
    }

    static func transcribe(
        fileURL: URL,
        context: String = "",
        screenContext: String = "",
        cleanup: Bool = true,
        completion: @escaping (Result<TranscriptionResult, Error>) -> Void
    ) {
        guard var components = URLComponents(url: baseURL.appendingPathComponent("transcribe"), resolvingAgainstBaseURL: false) else {
            completion(.failure(TranscriptionError.badResponse))
            return
        }
        components.queryItems = [URLQueryItem(name: "cleanup", value: cleanup ? "true" : "false")]

        guard let url = components.url, let audioData = try? Data(contentsOf: fileURL) else {
            completion(.failure(TranscriptionError.badResponse))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30

        // The sidecar is what talks to the cleanup gateway, so this Mac's device
        // key has to travel with the request. Sending it per call rather than
        // baking it into the sidecar's environment means re-pairing takes effect
        // immediately, with no restart, and the key is never written to disk
        // outside the Keychain.
        if let deviceKey = Keychain.deviceKey() {
            request.setValue("Bearer \(deviceKey)", forHTTPHeaderField: "X-SunoFlow-Device-Key")
        }

        let boundary = "Boundary-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"audio.wav\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: audio/wav\r\n\r\n".data(using: .utf8)!)
        body.append(audioData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"context\"\r\n\r\n".data(using: .utf8)!)
        body.append(context.data(using: .utf8) ?? Data())
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"screen\"\r\n\r\n".data(using: .utf8)!)
        body.append(screenContext.data(using: .utf8) ?? Data())
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let data = data else {
                completion(.failure(TranscriptionError.badResponse))
                return
            }
            // 402 means this Mac may not dictate. Deliberately not a soft
            // failure: nothing is pasted, and the user is told why.
            if (response as? HTTPURLResponse)?.statusCode == 402 {
                let body = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
                let message = body?["message"] as? String
                    ?? "Your SunoFlow subscription isn't active."
                let code = body?["error"] as? String ?? "not_entitled"
                completion(.failure(TranscriptionError.notEntitled(code: code, message: message)))
                return
            }
            do {
                let result = try JSONDecoder().decode(TranscriptionResult.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }

    // MARK: - Learning system

    private static func post(_ path: String, fields: [String: String], completion: ((Data?) -> Void)? = nil) {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.timeoutInterval = 5
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = formURLEncoded(fields)
        URLSession.shared.dataTask(with: request) { data, _, _ in
            completion?(data)
        }.resume()
    }

    /// Send a (pasted, edited) pair so the sidecar can learn corrections.
    static func learn(original: String, edited: String, completion: @escaping (Int) -> Void) {
        post("learn", fields: ["original": original, "edited": edited]) { data in
            guard let data = data,
                  let resp = try? JSONDecoder().decode(LearnResponse.self, from: data) else {
                completion(0)
                return
            }
            completion(resp.learned.count)
        }
    }

    static func fetchCorrections(completion: @escaping ([Correction]) -> Void) {
        var request = URLRequest(url: baseURL.appendingPathComponent("corrections"))
        request.timeoutInterval = 3
        URLSession.shared.dataTask(with: request) { data, _, _ in
            guard let data = data,
                  let resp = try? JSONDecoder().decode(CorrectionsResponse.self, from: data) else {
                completion([])
                return
            }
            completion(resp.corrections)
        }.resume()
    }

    static func deleteCorrection(key: String, completion: @escaping () -> Void) {
        post("corrections/delete", fields: ["key": key]) { _ in completion() }
    }

    static func clearCorrections(completion: @escaping () -> Void) {
        post("corrections/clear", fields: [:]) { _ in completion() }
    }

    /// Manually add an entry (from/to pair). The completion receives the updated
    /// full corrections list. The wire field is `frm` (not `from`) because
    /// `from` is a Python keyword and the FastAPI endpoint names the parameter
    /// `frm` — sending `from` gets a silent 422.
    ///
    /// A nil `kind` sends no `kind` field, leaving the sidecar to infer it from
    /// the shape of the pair. That is the default: the classifier lives in one
    /// place rather than being second-guessed here, and the list refreshes
    /// straight after with the badge showing what it decided.
    static func addCorrection(from: String, to: String, kind: CorrectionKind? = nil,
                              completion: @escaping ([Correction]) -> Void) {
        var fields = ["frm": from, "to": to]
        if let kind { fields["kind"] = kind.rawValue }
        post("corrections/add", fields: fields) { data in
            completion(Self.decodeCorrections(data))
        }
    }

    /// Edit an existing entry's from/to text. Wire field is `frm` (see above).
    /// The kind is always sent here: an edit should not silently reclassify an
    /// entry the user already placed.
    static func updateCorrection(key: String, from: String, to: String, kind: CorrectionKind? = nil,
                                 completion: @escaping ([Correction]) -> Void) {
        var fields = ["key": key, "frm": from, "to": to]
        if let kind { fields["kind"] = kind.rawValue }
        post("corrections/update", fields: fields) { data in
            completion(Self.decodeCorrections(data))
        }
    }

    private static func decodeCorrections(_ data: Data?) -> [Correction] {
        guard let data = data,
              let resp = try? JSONDecoder().decode(CorrectionsResponse.self, from: data) else {
            return []
        }
        return resp.corrections
    }

    // MARK: - Hosted cleanup gateway

    /// Reachability of the hosted cleanup service. Replaces the old local-Ollama
    /// probe — cleanup no longer runs on the user's machine. The gateway's
    /// `/ready` is unauthenticated and reports whether its LLM backend is up.
    static func checkCleanupGateway(completion: @escaping (Bool) -> Void) {
        guard let url = URL(string: gatewayURL.appendingPathComponent("ready").absoluteString) else {
            completion(false)
            return
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 4
        URLSession.shared.dataTask(with: request) { data, _, _ in
            // `/ready` returns {"backend_ok": true|false}. Treat any 200 with a
            // truthy backend_ok as reachable; fall back to false otherwise so the
            // UI shows "offline" rather than a misleading green dot.
            let ok = data
                .flatMap { try? JSONDecoder().decode(ReadyResponse.self, from: $0) }
                .map { $0.backend_ok } ?? false
            completion(ok)
        }.resume()
    }

    // MARK: - STT model download

    /// Current state of the Parakeet model (present on disk, loaded, download progress).
    static func fetchModelStatus(completion: @escaping (ModelStatus?) -> Void) {
        var request = URLRequest(url: baseURL.appendingPathComponent("model/status"))
        request.timeoutInterval = 4
        URLSession.shared.dataTask(with: request) { data, _, _ in
            let status = data.flatMap { try? JSONDecoder().decode(ModelStatus.self, from: $0) }
            completion(status)
        }.resume()
    }

    /// Kick off the background model download. Returns whether it actually
    /// started (false if one is already running or the model is already present).
    static func startModelDownload(completion: @escaping (Bool) -> Void) {
        var request = URLRequest(url: baseURL.appendingPathComponent("model/download"))
        request.httpMethod = "POST"
        request.timeoutInterval = 5
        URLSession.shared.dataTask(with: request) { data, _, _ in
            let started = data
                .flatMap { try? JSONDecoder().decode(DownloadStartResponse.self, from: $0) }
                .map { $0.started } ?? false
            completion(started)
        }.resume()
    }
}

File: apps/windows/SunoFlow/TranscriptionClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace SunoFlow;

/// <summary>
/// HTTP client for the SunoFlow sidecar (<c>127.0.0.1:8765</c>). This is the C#
/// counterpart of <c>TranscriptionClient.swift</c> and consumes the exact same
/// HTTP contract documented in <c>docs/CONTRACT.md</c>. Two notable fixes over
/// the Swift client, called out in the contract:
/// <list type="bullet">
/// <item><c>/corrections/add</c> + <c>/corrections/update</c> send <c>frm</c>
///   (not <c>from</c>) — <c>from</c> is a Python keyword, so the endpoint names
///   its parameter <c>frm</c> and anything else silently 422s.</item>
/// <item><c>/learn</c> decodes <c>learned</c> as a lighter struct with a
///   <c>count</c> field (the Swift client decodes into a struct requiring
///   <c>key</c> the server never sends, so the learned count always reads 0).</item>
/// </list>
/// </summary>
internal static class TranscriptionClient
{
    public const string BaseUrl = "http://127.0.0.1:8765";

    /// <summary>Hosted cleanup gateway (transcript polishing). The app only
    /// probes its unauthenticated <c>/ready</c> for the settings status card.</summary>
    public const string GatewayUrl = "http://162.19.81.108:40009";

    private static readonly HttpClient Http = new()
    {
        Timeout = TimeSpan.FromSeconds(30),
    };

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    // --- Response DTOs -------------------------------------------------------------

    public sealed class TranscriptionResult
    {
        [JsonPropertyName("raw")] public string Raw { get; set; } = "";
        [JsonPropertyName("cleaned")] public string Cleaned { get; set; } = "";
    }

    /// <summary>Which half of the dictionary an entry belongs to.
    ///
    /// The two behave very differently at dictation time — a spelling is swapped
    /// in mechanically, a shorthand value is only substituted where the model
    /// judges the speaker was actually giving it — so the UI names the
    /// difference rather than showing one undifferentiated list. See
    /// <c>docs/CONTRACT.md</c> §corrections.</summary>
    public enum CorrectionKind { Correction, Expansion }

    /// <summary>User-facing wording for a <see cref="CorrectionKind"/>, kept
    /// beside the enum so the two platforms stay worded the same.</summary>
    public static class CorrectionKinds
    {
        public static readonly CorrectionKind[] All =
            { CorrectionKind.Correction, CorrectionKind.Expansion };

        /// <summary>Anything unrecognised — including the field being absent,
        /// which is how a sidecar predating the two-kind dictionary answers —
        /// reads as a spelling. Those installs only ever held spellings.</summary>
        public static CorrectionKind Parse(string? wire) =>
            string.Equals(wire, "expansion", StringComparison.OrdinalIgnoreCase)
                ? CorrectionKind.Expansion
                : CorrectionKind.Correction;

        public static string Wire(CorrectionKind kind) =>
            kind == CorrectionKind.Expansion ? "expansion" : "correction";

        public static string Label(CorrectionKind kind) =>
            kind == CorrectionKind.Expansion ? "Shorthand" : "Spelling";

        public static string Blurb(CorrectionKind kind) => kind == CorrectionKind.Expansion
            ? "Something you say out loud, and the value it stands for."
            : "A word dictation mishears, and how you write it.";

        public static string FromPlaceholder(CorrectionKind kind) =>
            kind == CorrectionKind.Expansion ? "What you say" : "What it mishears";

        public static string ToPlaceholder(CorrectionKind kind) =>
            kind == CorrectionKind.Expansion ? "The value to write instead" : "What it should write";
    }

    public sealed class Correction
    {
        [JsonPropertyName("key")] public string Key { get; set; } = "";
        [JsonPropertyName("from")] public string From { get; set; } = "";
        [JsonPropertyName("to")] public string To { get; set; } = "";
        [JsonPropertyName("count")] public int Count { get; set; }
        /// <summary>Kept as the raw wire string so an unknown value can never
        /// break deserialization of the whole list; read it via
        /// <see cref="ResolvedKind"/>.</summary>
        [JsonPropertyName("kind")] public string Kind { get; set; } = "";

        [JsonIgnore] public CorrectionKind ResolvedKind => CorrectionKinds.Parse(Kind);
    }

    private sealed class CorrectionsResponse
    {
        [JsonPropertyName("corrections")] public Correction[] Corrections { get; set; } = Array.Empty<Correction>();
    }

    private sealed class LearnResponse
    {
        // Decoded as a lighter struct — see class doc.
        public sealed class LearnedItem
        {
            [JsonPropertyName("from")] public string From { get; set; } = "";
            [JsonPropertyName("to")] public string To { get; set; } = "";
            [JsonPropertyName("count")] public int Count { get; set; }
        }
        [JsonPropertyName("learned")] public LearnedItem[] Learned { get; set; } = Array.Empty<LearnedItem>();
        [JsonPropertyName("total")] public int Total { get; set; }
    }

    private sealed class ReadyResponse
    {
        [JsonPropertyName("backend_ok")] public bool BackendOk { get; set; }
    }

    /// <summary>The sidecar's 402 body when this device may not dictate.</summary>
    private sealed class RefusalResponse
    {
        [JsonPropertyName("error")] public string Error { get; set; } = "";
        [JsonPropertyName("message")] public string Message { get; set; } = "";
    }

    /// <summary>
    /// What came back from <c>/transcribe</c>. Three outcomes, and the caller
    /// has to tell them apart: a transcript, a refusal (the account's trial has
    /// ended, its subscription lapsed, or this PC is not connected), or a plain
    /// failure. A refusal must not be treated as a failure — the user needs to
    /// be told why nothing was typed, and pointed at the fix.
    /// </summary>
    public sealed class TranscribeOutcome
    {
        public TranscriptionResult? Result { get; init; }

        /// <summary>Non-null when the device was refused; carries the wording the
        /// gateway chose, so every surface shows one consistent explanation.</summary>
        public string? NotEntitled { get; init; }

        /// <summary>Why it was refused: <c>not_entitled</c> (trial over,
        /// subscription lapsed, device disconnected — fixed on the account
        /// page) or <c>unreachable</c> (we could not check — fixed by
        /// reconnecting). Both stop the dictation; they are not the same news
        /// and must not read the same in the tray.</summary>
        public string NotEntitledCode { get; init; } = "not_entitled";

        public static TranscribeOutcome Ok(TranscriptionResult r) => new() { Result = r };
        public static TranscribeOutcome Refused(string message, string code) =>
            new() { NotEntitled = message, NotEntitledCode = code };
        public static TranscribeOutcome Failed() => new();
    }

    public sealed class Health
    {
        [JsonPropertyName("status")] public string Status { get; set; } = "";
        [JsonPropertyName("model_loaded")] public bool ModelLoaded { get; set; }
        [JsonPropertyName("model_present")] public bool ModelPresent { get; set; }

        /// <summary>Why the model would not start, empty when it started or was
        /// never downloaded. Carried on the liveness probe so the tray — which
        /// polls nothing else — can stop telling someone to download a model
        /// that is already on their disk.</summary>
        [JsonPropertyName("load_error")] public string LoadError { get; set; } = "";
    }

    public sealed class ModelStatus
    {
        [JsonPropertyName("model_present")] public bool ModelPresent { get; set; }
        [JsonPropertyName("model_loaded")] public bool ModelLoaded { get; set; }
        [JsonPropertyName("active")] public bool Active { get; set; }
        [JsonPropertyName("phase")] public string Phase { get; set; } = "idle";
        [JsonPropertyName("current_file")] public string CurrentFile { get; set; } = "";
        [JsonPropertyName("downloaded")] public long Downloaded { get; set; }
        [JsonPropertyName("file_total")] public long FileTotal { get; set; }
        [JsonPropertyName("overall_done")] public int OverallDone { get; set; }
        [JsonPropertyName("overall_total")] public int OverallTotal { get; set; }
        /// <summary>Last <b>download</b> error. Its remedy is to download again.</summary>
        [JsonPropertyName("error")] public string Error { get; set; } = "";
        [JsonPropertyName("model_dir")] public string ModelDir { get; set; } = "";
        [JsonPropertyName("model_id")] public string ModelId { get; set; } = "";

        /// <summary>Last <b>load</b> error — a distinct failure from
        /// <see cref="Error"/>, and one that re-downloading cannot fix: by the
        /// time it happens the whole model is already on disk and intact enough
        /// to open.
        /// Empty when the model loaded, or was never downloaded.</summary>
        [JsonPropertyName("load_error")] public string LoadError { get; set; } = "";

        /// <summary>What inference was actually verified on — "GPU (DirectML)"
        /// or "CPU". The sidecar falls back to CPU on a box with no DX12 GPU,
        /// so the dashboard reads this rather than claiming the GPU.</summary>
        [JsonPropertyName("runtime")] public string Runtime { get; set; } = "";

        /// <summary>Which build of the model this PC runs — "fp32" or "int8".
        /// The sidecar picks it from the hardware it finds; it is not a setting,
        /// and there is no user choice to offer. Empty on a platform that ships
        /// only one build.</summary>
        [JsonPropertyName("variant")] public string Variant { get; set; } = "";

        /// <summary><see cref="Variant"/> in words, for display: "full precision"
        /// or "int8".</summary>
        [JsonPropertyName("variant_label")] public string VariantLabel { get; set; } = "";

        /// <summary>Why the sidecar chose that build, in a sentence a user can
        /// read. Worth showing: somebody told they are on the smaller model is
        /// owed the reason, and "your GPU has 2 GB" is a better answer than
        /// silence.</summary>
        [JsonPropertyName("variant_reason")] public string VariantReason { get; set; } = "";

        /// <summary>What this PC's variant weighs. Full precision is ~2.5 GB and
        /// int8 about a quarter of that, so no screen may quote a fixed size —
        /// read it from here via <see cref="DownloadSizeText"/>.</summary>
        [JsonPropertyName("download_bytes")] public long DownloadBytes { get; set; }

        /// <summary>The download size as a short phrase for prose. Falls back to
        /// a size-free wording before the sidecar has said which variant this PC
        /// gets, which is better than naming a number that may be four times
        /// too big.</summary>
        [JsonIgnore]
        public string DownloadSizeText =>
            DownloadBytes >= 1_000_000_000 ? $"about {DownloadBytes / 1_000_000_000.0:0.0} GB"
            : DownloadBytes >= 1_000_000 ? $"about {DownloadBytes / 1_000_000.0:0} MB"
            : "a one-time download";

        /// <summary>True when the model is downloaded but will not start. The
        /// one state the dashboard used to have no words for: it looked like
        /// "not downloaded" and was offered a fix aimed at something else.</summary>
        [JsonIgnore] public bool LoadFailed => !ModelLoaded && !string.IsNullOrWhiteSpace(LoadError);
    }

    // --- Health -------------------------------------------------------------------

    /// <summary>GET /health. Liveness plus the model flags, because "the engine
    /// answers" and "dictation can produce words" are different questions — a
    /// sidecar with no model downloaded is healthy and returns empty
    /// transcripts forever.</summary>
    public static async Task<Health?> HealthDetailAsync()
    {
        try
        {
            using var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(2));
            using var resp = await Http.GetAsync($"{BaseUrl}/health", cts.Token);
            if (!resp.IsSuccessStatusCode) return null;
            var json = await resp.Content.ReadAsStringAsync(cts.Token);
            return JsonSerializer.Deserialize<Health>(json, JsonOpts);
        }
        catch { return null; }
    }

    /// <summary>Lightweight liveness probe (GET /health). Polled every 3s.</summary>
    public static async Task<bool> HealthAsync() => await HealthDetailAsync() != null;

    // --- Transcribe (multipart) ---------------------------------------------------

    /// <summary>
    /// POST /transcribe. Uploads the WAV as multipart/form-data with the
    /// optional <c>context</c> and <c>screen</c> fields, plus the <c>cleanup</c>
    /// query param.
    ///
    /// This PC's device key travels with the request so the sidecar can
    /// authenticate to the cleanup gateway, which is where the account's trial
    /// and subscription are actually checked. Sending it per call rather than
    /// baking it into the sidecar's environment means re-pairing takes effect
    /// immediately, with no restart, and the key is never written to disk
    /// outside its DPAPI-protected store.
    /// </summary>
    public static async Task<TranscribeOutcome> TranscribeAsync(
        string wavPath, string context, string screenContext, bool cleanup)
    {
        try
        {
            using var form = new MultipartFormDataContent();
            var fileBytes = await File.ReadAllBytesAsync(wavPath);
            var fileContent = new ByteArrayContent(fileBytes);
            fileContent.Headers.ContentType = new MediaTypeHeaderValue("audio/wav");
            form.Add(fileContent, "file", "audio.wav");
            form.Add(new StringContent(context, Encoding.UTF8), "context");
            form.Add(new StringContent(screenContext, Encoding.UTF8), "screen");

            var url = $"{BaseUrl}/transcribe?cleanup={(cleanup ? "true" : "false")}";
            using var request = new HttpRequestMessage(HttpMethod.Post, url) { Content = form };
            var deviceKey = AccountManager.Shared.DeviceKey;
            if (!string.IsNullOrEmpty(deviceKey))
                request.Headers.TryAddWithoutValidation("X-SunoFlow-Device-Key", $"Bearer {deviceKey}");

            using var resp = await Http.SendAsync(request);

            // 402 means this device may not dictate: the trial ended, the
            // subscription lapsed, or the account disconnected this PC. This is
            // deliberately not a soft failure — nothing is pasted and the user
            // is told why.
            if ((int)resp.StatusCode == 402)
            {
                var body = await resp.Content.ReadAsStringAsync();
                string message = "Your SunoFlow subscription isn't active.";
                string code = "not_entitled";
                try
                {
                    var parsed = JsonSerializer.Deserialize<RefusalResponse>(body, JsonOpts);
                    if (!string.IsNullOrWhiteSpace(parsed?.Message)) message = parsed!.Message;
                    if (!string.IsNullOrWhiteSpace(parsed?.Error)) code = parsed!.Error;
                }
                catch { /* keep the default wording */ }
                AppLog.Log($"Dictation refused — {code}");
                return TranscribeOutcome.Refused(message, code);
            }

            if (!resp.IsSuccessStatusCode)
            {
                AppLog.Log($"Transcribe returned HTTP {resp.StatusCode}");
                return TranscribeOutcome.Failed();
            }
            var json = await resp.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<TranscriptionResult>(json, JsonOpts);
            return result == null ? TranscribeOutcome.Failed() : TranscribeOutcome.Ok(result);
        }
        catch (Exception ex)
        {
            AppLog.Log($"Transcribe request failed: {ex.Message}");
            return TranscribeOutcome.Failed();
        }
    }

    // --- Learning -----------------------------------------------------------------

    /// <summary>Send a (pasted, edited) pair so the sidecar can learn corrections.</summary>
    public static async Task<int> LearnAsync(string original, string edited)
    {
        var fields = new Dictionary<string, string> { ["original"] = original, ["edited"] = edited };
        var data = await PostFormAsync("learn", fields);
        if (data == null) return 0;
        try
        {
            var resp = JsonSerializer.Deserialize<LearnResponse>(data, JsonOpts);
            return resp?.Learned.Length ?? 0;
        }
        catch { return 0; }
    }

    public static async Task<Correction[]> FetchCorrectionsAsync()
    {
        try
        {
            using var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(3));
            var resp = await Http.GetAsync($"{BaseUrl}/corrections", cts.Token);
            if (!resp.IsSuccessStatusCode) return Array.Empty<Correction>();
            var json = await resp.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<CorrectionsResponse>(json, JsonOpts)?.Corrections
                   ?? Array.Empty<Correction>();
        }
        catch { return Array.Empty<Correction>(); }
    }

    public static async Task DeleteCorrectionAsync(string key)
        => await PostFormAsync("corrections/delete", new Dictionary<string, string> { ["key"] = key });

    public static async Task ClearCorrectionsAsync()
        => await PostFormAsync("corrections/clear", new Dictionary<string, string>());

    /// <summary>
    /// Manually add a correction. Sends <c>frm</c> on the wire (not
    /// <c>from</c>) — see the contract note. Returns the updated list.
    /// </summary>
    /// <param name="kind">Null sends no <c>kind</c> field, leaving the sidecar to
    /// infer it from the shape of the pair. That is the default: the classifier
    /// lives in one place rather than being second-guessed here, and the list
    /// refreshes straight after with the badge showing what it decided.</param>
    public static async Task<Correction[]> AddCorrectionAsync(string from, string to,
                                                              CorrectionKind? kind = null)
    {
        var fields = new Dictionary<string, string> { ["frm"] = from, ["to"] = to };
        if (kind.HasValue) fields["kind"] = CorrectionKinds.Wire(kind.Value);
        var data = await PostFormAsync("corrections/add", fields);
        return DecodeCorrections(data);
    }

    /// <summary>Edit an existing correction. Sends <c>frm</c> + <c>key</c>.</summary>
    /// <param name="kind">Always worth sending on an edit: fixing a typo in a URL
    /// should not silently reclassify the entry.</param>
    public static async Task<Correction[]> UpdateCorrectionAsync(string key, string from, string to,
                                                                 CorrectionKind? kind = null)
    {
        var fields = new Dictionary<string, string> { ["key"] = key, ["frm"] = from, ["to"] = to };
        if (kind.HasValue) fields["kind"] = CorrectionKinds.Wire(kind.Value);
        var data = await PostFormAsync("corrections/update", fields);
        return DecodeCorrections(data);
    }

    private static Correction[] DecodeCorrections(string? data)
    {
        if (data == null) return Array.Empty<Correction>();
        try
        {
            return JsonSerializer.Deserialize<CorrectionsResponse>(data, JsonOpts)?.Corrections
                   ?? Array.Empty<Correction>();
        }
        catch { return Array.Empty<Correction>(); }
    }

    // --- Cleanup gateway reachability --------------------------------------------

    /// <summary>Reachability of the hosted cleanup service (GET /ready, no auth).</summary>
    public static async Task<bool> CheckCleanupGatewayAsync()
    {
        try
        {
            using var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(4));
            var resp = await Http.GetAsync($"{GatewayUrl}/ready", cts.Token);
            if (!resp.IsSuccessStatusCode) return false;
            var json = await resp.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<ReadyResponse>(json, JsonOpts)?.BackendOk ?? false;
        }
        catch { return false; }
    }

    // --- Model download -----------------------------------------------------------

    public static async Task<ModelStatus?> FetchModelStatusAsync()
    {
        try
        {
            using var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(4));
            var resp = await Http.GetAsync($"{BaseUrl}/model/status", cts.Token);
            if (!resp.IsSuccessStatusCode) return null;
            var json = await resp.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<ModelStatus>(json, JsonOpts);
        }
        catch { return null; }
    }

    public static async Task<bool> StartModelDownloadAsync()
    {
        try
        {
            using var resp = await Http.PostAsync($"{BaseUrl}/model/download", new StringContent(""));
            if (!resp.IsSuccessStatusCode) return false;
            var json = await resp.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.TryGetProperty("started", out var s) && s.GetBoolean();
        }
        catch { return false; }
    }

    // --- Internal helper ---------------------------------------------------------

    private static async Task<string?> PostFormAsync(string path, Dictionary<string, string> fields)
    {
        try
        {
            using var cts = new System.Threading.CancellationTokenSource(TimeSpan.FromSeconds(5));
            var content = new FormUrlEncodedContent(fields);
            var resp = await Http.PostAsync($"{BaseUrl}/{path}", content, cts.Token);
            return await resp.Content.ReadAsStringAsync();
        }
        catch (Exception ex)
        {
            AppLog.Log($"POST /{path} failed: {ex.Message}");
            return null;
        }
    }
}